### PR TITLE
Add support for bucket replication

### PR DESCRIPTION
**Pull request by aminehmida**
I split this commit to make my changes easier to check. Please have a look at https://github.com/mineiros-io/terraform-aws-s3-bucket/commit/8fe3fe3b91b414548760697543bdd8540fddd834 and let me know if anything need to be done to merge this in master.

Thanks for this awesome module.

### CODE REVIEW

**Comment by mariux** [comment]
Hi @aminehmida, 

thanks for your great contribution. I will discuss needed changes to your PR with the team and add comments within the next days. 

We do have some internal guidelines to keep base modules within one service and not add other services like roles into them.. but it makes totally sense to create the replication role with the replication configuration tbh. 

so far:

> I split this commit to make my changes easier to check. Please have a look at [8fe3fe3](https://github.com/mineiros-io/terraform-aws-s3-bucket/commit/8fe3fe3b91b414548760697543bdd8540fddd834) and let me know if anything need to be done to merge this in master.

no need to split commits here, as you can disable whitespace changes in the review console of github.. - check it out now ;) could you squash the two commits and force push again? not a hard requirement though ;)

Thanks again :rocket:

File: main.tf
Comment (mariux) [inline main.tf:437]: according to the example in the provider this should be:

```suggestion
        {
          "Action" : [
            "s3:GetReplicationConfiguration",
            "s3:ListBucket",
          ],
          "Effect" : "Allow",
          "Resource" : [
            "${local.bucket_arn}",
          ]
        },
        {
          "Action" : [
            "s3:GetObjectVersion",
            "s3:GetObjectVersionAcl",
            "s3:GetObjectVersionTagging",
          ],
          "Effect" : "Allow",
          "Resource" : [
            "${local.bucket_arn}/*"
          ]
        },
```

File: main.tf
Comment (mariux) [inline main.tf:451]: according to the example in the provider this should be:

also https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication-walkthrough1.html specifies the same..

where are your policies taken from?

```suggestion
        {
          "Action" : [
            "s3:ReplicateObject",
            "s3:ReplicateDelete",
            "s3:ReplicateTags",
          ],
          "Effect" : "Allow",
          "Resource" : [
            "arn:aws:s3:::${var.replication.bucket}/*",
          ]
        }
```

File: main.tf
Comment (mariux) [inline main.tf:419]: please use `aws_iam_policy_document` data source instead of `jsonencode` for *all* policies as this fixes some AWS specific issues to prevent policies from constantly showing changes at plan time.

File: main.tf
Comment (mariux) [inline main.tf:396]: this should be given by the user.

```suggestion
  name       = try(var.replication.role_name, "${var.bucket}-replication-role")
```

File: main.tf
Comment (mariux) [inline main.tf:457]: To reduce number of resources to refresh we should make this an inline policy in the role and not create an extra policy that will never be reused.

File: main.tf
Comment (mariux) [inline main.tf:165]: this block is optional according to the documentation.

File: main.tf
Comment (mariux) [inline main.tf:156]: should be set by rule i assume as we want to support more than one rule?

**Review by mariux — CHANGES_REQUESTED** [review]
soo... no real content or logic changes here... awesome work.. thanks a lot.. some blocks can be set multiple times according to the source code of the provider... we should support this also in the module..

as mentioned in the other comment we need to discuss creating a role in a bucket module. for me this makes totally sense here.. 

once the PR is ready to be merged we should also update the readme ;) but let's get the details right first ...


cc @soerenmartius

File: main.tf
Comment (samen93) [reply main.tf:156]: can you replace lookup with try() too?

**Review by samen93 — CHANGES_REQUESTED** [review]

**Review by samen93 — COMMENTED** [review]
will review

**Review by samen93 — COMMENTED** [review]
looks good to me

**Comment by Limess** [comment]
Hey, anything blocking this from being merged? We'd love to see these changes in master.

**Comment by soerenmartius** [comment]
> Hey, anything blocking this from being merged? We'd love to see these changes in master.

Hi @Limess , 

Let me discuss this with the team. This PR needs some changes in order to be merged back to master but we might be able to handle this internally this week.

**Comment by aminehmida** [comment]
This is very old. I am not working on this anymore. Feel free to use my code and make another PR if you like. Just going to close it for now.

**Pull request closed without merging**
